drv/lpc55-swd/src/main.rs

```diff
@@ -0,0 +1,656 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use drv_lpc55_spi as spi_core;
+use drv_lpc55_syscon_api::{Peripheral, Syscon};
+use drv_sp_ctrl_api::SpCtrlError;
+use idol_runtime::{
+    LeaseBufReader, LeaseBufWriter, Leased, LenLimit, RequestError, R, W,
+};
+use lpc55_pac as device;
+use ringbuf::*;
+use userlib::*;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    Idcode(u32),
+    Idr(u32),
+    MemVal(u32),
+    SwdRead(Port, RawSwdReg),
+    SwdWrite(Port, RawSwdReg, u32),
+    ReadCmd,
+    WriteCmd,
+    None,
+    AckErr(Ack),
+}
+
+ringbuf!(Trace, 64, Trace::None);
+
+task_slot!(SYSCON, syscon_driver);
+task_slot!(GPIO, gpio_driver);
+
+#[derive(Copy, Clone, PartialEq)]
+enum Ack {
+    //Ok,
+    Wait,
+    Fault,
+    Protocol,
+}
+
+// Several of these bits in the CSW are  implementation defined according to
+// the official specification but thankfully RM0433 60.4.2 actually gives us the
+// defintions
+
+// Full 32-bit word transfer
+const CSW_SIZE32: u32 = 0x00000002;
+// Increment by size bytes in the transaction
+const CSW_SADDRINC: u32 = 0x00000010;
+// AP access enabled
+const CSW_DBGSTAT: u32 = 0x00000040;
+// Privileged + data access
+const CSW_HPROT: u32 = 0x03000000;
+
+const DP_CTRL_CDBGPWRUPREQ: u32 = 1 << 28;
+const DP_CTRL_CDBGPWRUPACK: u32 = 1 << 29;
+
+// See Ch5 of ARM ADI for bit pattern
+const START_BIT: u8 = 7;
+// Stop is bit 1 and always 0
+const PARITY_BIT: u8 = 2;
+const ADDR_BITS: u8 = 3;
+
+const RDWR_BIT: u8 = 5;
+const APDP_BIT: u8 = 6;
+const PARK_BIT: u8 = 0;
+
+const START_VAL: u8 = 1 << START_BIT;
+const PARK_VAL: u8 = 1 << PARK_BIT;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Port {
+    DP = 0,
+    AP = 1,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq)]
+enum DpRead {
+    IDCode = 0x0,
+    Ctrl = 0x4,
+    //Resend = 0x8,
+    //Rdbuf = 0xc,
+}
+
+impl DpRead {
+    fn addr_bits(&self) -> u8 {
+        // we need this to be as bits [2:3]
```

Comment:
This comment and the implementation of this and DpWrite below are confusing me. Can you expand on it? I think you're deliberately bit-reversing?